File: src/odioctl/dac.py
"""DAC selection: manage the `dtoverlay=` line in the Raspberry Pi config.txt.

odioctl owns one marked block at the end of config.txt:

    # BEGIN odioctl dac -- managed block, edit with `odioctl dac`
    [all]
    dtparam=audio=off
    dtoverlay=hifiberry-dacplus-std
    # END odioctl dac

The `[all]` resets any `[pi4]`/`[cm5]`-style filter section that may be
open above so the block always applies. Pre-existing top-level audio lines
(`dtparam=audio=…`, `dtoverlay=<known DAC overlay>`) are commented out with
a recognisable prefix so `dac unset` can restore them verbatim. Text
manipulation is pure (`parse`/`apply`/`unapply`) and unit-tested; only
`read_config`/`write_config` touch the disk. Writing needs root — the web
UI and unprivileged callers go through `sudo odioctl dac set <id>`, and the
sudoers wildcard is safe because argparse `choices` re-validates the id.
"""

from __future__ import annotations

import argparse
import contextlib
import json
import os
import re
import shutil
import sys
from dataclasses import asdict, dataclass

from odioctl import fsutil

CONFIG_CANDIDATES = ("/boot/firmware/config.txt", "/boot/config.txt")
REBOOT_FLAG = "/run/odioctl/reboot-required"

BEGIN = "# BEGIN odioctl dac -- managed block, edit with `odioctl dac`"
END = "# END odioctl dac"
DISABLED_PREFIX = "#odioctl-disabled: "

ONBOARD = "onboard"


@dataclass(frozen=True)
class DacEntry:
    id: str
    label: str
    overlay: str | None
    params: str = ""

    def dtoverlay_line(self) -> str | None:
        if self.overlay is None:
            return None
        return f"dtoverlay={self.overlay}" + (f",{self.params}" if self.params else "")


CATALOG: tuple[DacEntry, ...] = (
    DacEntry(ONBOARD, "Onboard audio (3.5mm jack / HDMI)", None),
    # HiFiBerry
    DacEntry(
        "hifiberry-dac",
        "HiFiBerry DAC / DAC+ Light / DAC Zero / MiniAmp / PCM5102",
        "hifiberry-dac",
    ),
    DacEntry("hifiberry-dacplus-std", "HiFiBerry DAC+ (standard)", "hifiberry-dacplus-std"),
    DacEntry("hifiberry-dacplus-pro", "HiFiBerry DAC+ Pro / DAC2 Pro", "hifiberry-dacplus-pro"),
    DacEntry("hifiberry-dacplusadc", "HiFiBerry DAC+ ADC", "hifiberry-dacplusadc"),
    DacEntry(
        "hifiberry-dacplusadcpro",
        "HiFiBerry DAC+ ADC Pro / DAC2 ADC Pro",
        "hifiberry-dacplusadcpro",
    ),
    DacEntry("hifiberry-dacplushd", "HiFiBerry DAC+ HD / DAC2 HD", "hifiberry-dacplushd"),
    DacEntry("hifiberry-digi", "HiFiBerry Digi / Digi+", "hifiberry-digi"),
    DacEntry("hifiberry-digi-pro", "HiFiBerry Digi+ Pro / Digi2 Pro", "hifiberry-digi-pro"),
    DacEntry("hifiberry-amp", "HiFiBerry Amp / Amp+", "hifiberry-amp"),
    DacEntry("hifiberry-amp100", "HiFiBerry Amp100", "hifiberry-amp100"),
    DacEntry("hifiberry-amp3", "HiFiBerry Amp3", "hifiberry-amp3"),
    # IQaudIO / Raspberry Pi
    DacEntry("iqaudio-dac", "IQaudIO Pi-DAC / Pi-DAC Zero", "iqaudio-dac"),
    DacEntry(
        "iqaudio-dacplus",
        "IQaudIO Pi-DAC+ / Pi-DAC Pro / Pi-DigiAMP+ / Raspberry Pi DAC Pro",
        "iqaudio-dacplus",
    ),
    DacEntry(
        "iqaudio-digi-wm8804-audio",
        "IQaudIO Pi-Digi+ / Raspberry Pi DigiAMP+",
        "iqaudio-digi-wm8804-audio",
    ),
    DacEntry("iqaudio-codec", "IQaudIO Pi-Codec+ / Raspberry Pi Codec Zero", "iqaudio-codec"),
    DacEntry("rpi-codeczero", "Raspberry Pi Codec Zero (legacy overlay)", "rpi-codeczero"),
    # Allo
    DacEntry("allo-boss-dac-pcm512x-audio", "Allo Boss DAC", "allo-boss-dac-pcm512x-audio"),
    DacEntry("allo-boss2-dac-audio", "Allo Boss2 DAC", "allo-boss2-dac-audio"),
    DacEntry("allo-piano-dac-pcm512x-audio", "Allo Piano DAC", "allo-piano-dac-pcm512x-audio"),
    DacEntry(
        "allo-piano-dac-plus-pcm512x-audio",
        "Allo Piano DAC 2.1",
        "allo-piano-dac-plus-pcm512x-audio",
    ),
    DacEntry("allo-digione", "Allo DigiOne", "allo-digione"),
    DacEntry("allo-katana-dac-audio", "Allo Katana DAC", "allo-katana-dac-audio"),
    # JustBoom
    DacEntry("justboom-dac", "JustBoom DAC / Amp", "justboom-dac"),
    DacEntry("justboom-digi", "JustBoom Digi", "justboom-digi"),
    DacEntry("justboom-both", "JustBoom DAC + Digi", "justboom-both"),
    # Others
    DacEntry(
        "audioinjector-wm8731-audio", "AudioInjector Zero / Stereo", "audioinjector-wm8731-audio"
    ),
    DacEntry("pisound", "Blokas pisound", "pisound"),
    DacEntry("googlevoicehat-soundcard", "Google AIY Voice HAT", "googlevoicehat-soundcard"),
)

BY_ID: dict[str, DacEntry] = {e.id: e for e in CATALOG}
BY_OVERLAY: dict[str, DacEntry] = {e.overlay: e for e in CATALOG if e.overlay}
KNOWN_OVERLAYS = frozenset(BY_OVERLAY)
# Any overlay from these families counts as an audio line even when it is not in the
# catalog (legacy names such as `hifiberry-dacplus`, vendor variants): it is commented
# out by apply() and reported as stray instead of being left to fight the managed block.
AUDIO_OVERLAY_PREFIXES = (
    "hifiberry-",
    "iqaudio-",
    "rpi-codeczero",
    "allo-",
    "justboom-",
    "audioinjector-",
    "pisound",
    "googlevoicehat-",
    "i-sabre-",
    "dionaudio-",
    "fe-pi-",
    "rpi-dac",
    "rpi-proto",
    "udrc",
    "adau",
)

_OVERLAY_RE = re.compile(r"^\s*dtoverlay\s*=\s*([^,\s]+)(?:,(.*))?\s*$")
_AUDIO_RE = re.compile(r"^\s*dtparam\s*=\s*audio\s*=\s*(on|off)\s*$", re.IGNORECASE)


@dataclass
class DacStatus:
    current: str | None  # DacEntry.id, or None when nothing recognisable is configured
    managed: bool  # True when the odioctl block is present
    stray_lines: list[str]  # active audio lines outside the managed block


def _split_block(lines: list[str]) -> tuple[list[str], list[str] | None, list[str]]:
    """Return (before, block_lines_or_None, after). A BEGIN without END swallows to EOF."""
    try:
        start = lines.index(BEGIN)
    except ValueError:
        return lines, None, []
    try:
        end = lines.index(END, start + 1)
    except ValueError:
        end = len(lines) - 1
    return lines[:start], lines[start : end + 1], lines[end + 1 :]


def _entry_for_lines(lines: list[str]) -> str | None:
    audio_on = False
    for line in lines:
        m = _OVERLAY_RE.match(line)
        if m and m.group(1) in BY_OVERLAY:
            return BY_OVERLAY[m.group(1)].id
        a = _AUDIO_RE.match(line)
        if a and a.group(1).lower() == "on":
            audio_on = True
    return ONBOARD if audio_on else None


def is_audio_overlay(name: str) -> bool:
    return name in KNOWN_OVERLAYS or name.startswith(AUDIO_OVERLAY_PREFIXES)


def _is_audio_line(line: str) -> bool:
    m = _OVERLAY_RE.match(line)
    if m and is_audio_overlay(m.group(1)):
        return True
    return bool(_AUDIO_RE.match(line))


def parse(text: str) -> DacStatus:
    lines = text.splitlines()
    before, block, after = _split_block(lines)
    outside = before + after
    stray = [ln for ln in outside if _is_audio_line(ln)]
    if block is not None:
        return DacStatus(current=_entry_for_lines(block), managed=True, stray_lines=stray)
    return DacStatus(current=_entry_for_lines(outside), managed=False, stray_lines=stray)


def render_block(entry: DacEntry) -> list[str]:
    lines = [BEGIN, "[all]", f"dtparam=audio={'on' if entry.overlay is None else 'off'}"]
    dto = entry.dtoverlay_line()
    if dto:
        lines.append(dto)
    lines.append(END)
    return lines


def _strip_trailing_blank(lines: list[str]) -> list[str]:
    while lines and not lines[-1].strip():
        lines.pop()
    return lines


def apply(text: str, entry: DacEntry) -> str:
    """Return config.txt text with `entry` selected. Idempotent."""
    before, _block, after = _split_block(text.splitlines())
    body: list[str] = []
    for line in before + after:
        if _is_audio_line(line):
            body.append(DISABLED_PREFIX + line)
        else:
            body.append(line)
    body = _strip_trailing_blank(body)
    if body:
        body.append("")
    body.extend(render_block(entry))
    return "\n".join(body) + "\n"


def unapply(text: str) -> str:
    """Return config.txt text with the managed block removed and disabled lines restored."""
    before, _block, after = _split_block(text.splitlines())
    body: list[str] = []
    for line in before + after:
        if line.startswith(DISABLED_PREFIX):
            body.append(line[len(DISABLED_PREFIX) :])
        else:
            body.append(line)
    body = _strip_trailing_blank(body)
    return ("\n".join(body) + "\n") if body else ""


# --- I/O -------------------------------------------------------------------


def find_config_txt() -> str | None:
    for p in CONFIG_CANDIDATES:
        if os.path.isfile(p):
            return p
    return None


def read_config(path: str) -> str:
    with open(path, encoding="utf-8", errors="replace") as f:
        return f.read()


def write_config(path: str, text: str) -> None:
    backup = path + ".odioctl.bak"
    if not os.path.exists(backup):
        with contextlib.suppress(OSError):
            shutil.copy2(path, backup)
    fsutil.atomic_write_text(path, text)


def mark_reboot_required() -> None:
    with contextlib.suppress(OSError):
        os.makedirs(os.path.dirname(REBOOT_FLAG), exist_ok=True)
        with open(REBOOT_FLAG, "w") as f:
            f.write("dac\n")


def reboot_required() -> bool:
    return os.path.exists(REBOOT_FLAG)


def status(config_path: str | None = None) -> dict[str, object]:
    """Status dict shared by the CLI and the web API."""
    path = config_path or find_config_txt()
    if path is None or not os.path.isfile(path):
        return {
            "supported": False,
            "config": path,
            "current": None,
            "managed": False,
            "stray_lines": [],
            "reboot_required": reboot_required(),
        }
    st = parse(read_config(path))
    return {
        "supported": True,
        "config": path,
        "current": st.current,
        "managed": st.managed,
        "stray_lines": st.stray_lines,
        "reboot_required": reboot_required(),
    }


# --- CLI -----------------------------------------------------------------


def add_dac_arguments(p: argparse.ArgumentParser) -> None:
    p.description = "Select the DAC (dtoverlay) in the Raspberry Pi config.txt."
    sub = p.add_subparsers(dest="dac_cmd", metavar="COMMAND", required=True)
    ls = sub.add_parser("list", help="list supported DACs")
    ls.add_argument("--json", action="store_true")
    stt = sub.add_parser("status", help="show the currently configured DAC")
    stt.add_argument("--json", action="store_true")
    stt.add_argument("--config", help="path to config.txt (default: auto-detect)")
    st = sub.add_parser("set", help="select a DAC (root; reboot required)")
    st.add_argument("id", choices=sorted(BY_ID), metavar="ID", help="DAC id from `dac list`")
    st.add_argument("--config", help="path to config.txt (default: auto-detect)")
    st.add_argument("--dry-run", action="store_true", help="print the resulting file, don't write")
    un = sub.add_parser("unset", help="remove the odioctl block, restore previous lines (root)")
    un.add_argument("--config", help="path to config.txt (default: auto-detect)")
    un.add_argument("--dry-run", action="store_true")


def dac_from_args(ns: argparse.Namespace) -> int:
    if ns.dac_cmd == "list":
        if ns.json:
            print(json.dumps([asdict(e) for e in CATALOG], indent=2))
        else:
            for e in CATALOG:
                print(f"  {e.id:<36} {e.label}")
        return 0

    if ns.dac_cmd == "status":
        s = status(ns.config)
        if ns.json:
            print(json.dumps(s, indent=2))
        elif not s["supported"]:
            print("no config.txt found — not a Raspberry Pi boot partition?")
        else:
            cur = s["current"]
            label = BY_ID[str(cur)].label if isinstance(cur, str) and cur in BY_ID else "(unknown)"
            print(f"config:   {s['config']}")
            print(f"current:  {cur or '(none)'} {label if cur else ''}".rstrip())
            print(f"managed:  {'yes' if s['managed'] else 'no'}")
            if s["reboot_required"]:
                print("reboot required to apply the last change")
        return 0

    path = ns.config or find_config_txt()
    if path is None or not os.path.isfile(path):
        print(
            "Error: no config.txt found (tried " + ", ".join(CONFIG_CANDIDATES) + ")",
            file=sys.stderr,
        )
        return 2
    text = read_config(path)
    new = apply(text, BY_ID[ns.id]) if ns.dac_cmd == "set" else unapply(text)

    if ns.dry_run:
        sys.stdout.write(new)
        return 0
    if new == text:
        print("no change")
        return 0
    try:
        write_config(path, new)
    except OSError as e:
        print(f"Error writing {path}: {e}", file=sys.stderr)
        return 2
    mark_reboot_required()
    what = f"DAC set to {ns.id}" if ns.dac_cmd == "set" else "odioctl DAC block removed"
    print(f"{what} in {path} — reboot required")
    return 0

File: dac/dac.go
// Package dac manages the `dtoverlay=` line in the Raspberry Pi config.txt.
//
// odioctl owns one marked block at the end of config.txt:
//
//	# BEGIN odioctl dac -- managed block, edit with `odioctl dac`
//	[all]
//	dtoverlay=
//	dtparam=audio=off
//	dtoverlay=hifiberry-dacplus-std
//	# END odioctl dac
//
// The [all] resets any open [pi4]-style filter section, and the empty
// `dtoverlay=` retargets `dtparam=` at the base DTB (see overlayReset).
// Pre-existing top-level audio lines are commented out with DisabledPrefix so
// `dac unset` restores them verbatim. Parse/Apply/Unapply are pure; only
// ReadConfig/WriteConfig touch the disk. Writing needs root — unprivileged
// callers go through `sudo odioctl dac set <id>`.
package dac

import (
	"log/slog"
	"os"
	"path/filepath"
	"regexp"
	"slices"
	"strings"

	"github.com/b0bbywan/odioctl/fsutil"
)

var ConfigCandidates = []string{"/boot/firmware/config.txt", "/boot/config.txt"}

// RebootFlag is a var so tests can point it into a tempdir.
var RebootFlag = "/run/odioctl/reboot-required"

const (
	Begin = "# BEGIN odioctl dac -- managed block, edit with `odioctl dac`"
	End   = "# END odioctl dac"
	// An empty `dtoverlay=` ends the parameter list of the overlay above it:
	// without this reset our `dtparam=audio=` would land on whatever the file
	// loaded last instead of the base DTB, and snd_bcm2835 never loads.
	overlayReset   = "dtoverlay="
	DisabledPrefix = "#odioctl-disabled: "

	Onboard = "onboard"
)

type Entry struct {
	ID     string `json:"id"` // also the dtoverlay name
	Label  string `json:"label"`
	Params string `json:"params"`
}

// OverlayLine is the config.txt line, "" for onboard audio (no overlay).
func (e Entry) OverlayLine() string {
	if e.ID == Onboard {
		return ""
	}
	if e.Params != "" {
		return "dtoverlay=" + e.ID + "," + e.Params
	}
	return "dtoverlay=" + e.ID
}

var Catalog = []Entry{
	{ID: Onboard, Label: "Onboard audio (3.5mm jack / HDMI)"},
	// Generic I2S
	{ID: "i2s-dac", Label: "Generic passive I2S DAC (Pi as clock master)"},
	{ID: "i2s-master-dac", Label: "Generic I2S DAC acting as clock master"},
	// HiFiBerry
	{ID: "hifiberry-dac", Label: "HiFiBerry DAC / DAC+ Light / DAC+ Zero / MiniAmp / PCM5102"},
	{ID: "hifiberry-dacplus", Label: "HiFiBerry DAC+ (auto-detect std/pro)"},
	{ID: "hifiberry-dacplus-std", Label: "HiFiBerry DAC+ (standard)"},
	{ID: "hifiberry-dacplus-pro", Label: "HiFiBerry DAC+ Pro / DAC2 Pro"},
	{ID: "hifiberry-dacplusadc", Label: "HiFiBerry DAC+ ADC"},
	{ID: "hifiberry-dacplusadcpro", Label: "HiFiBerry DAC+ ADC Pro / DAC2 ADC Pro"},
	{ID: "hifiberry-dacplushd", Label: "HiFiBerry DAC+ HD / DAC2 HD"},
	{ID: "hifiberry-dacplusdsp", Label: "HiFiBerry DAC+ DSP"},
	{ID: "hifiberry-dac8x", Label: "HiFiBerry DAC8x (Pi 5 only)"},
	{ID: "hifiberry-studio-dac8x", Label: "HiFiBerry Studio DAC8x"},
	{ID: "hifiberry-studio-dac8x-pro", Label: "HiFiBerry Studio DAC8x Pro"},
	{ID: "hifiberry-digi", Label: "HiFiBerry Digi / Digi+"},
	{ID: "hifiberry-digi-pro", Label: "HiFiBerry Digi+ Pro / Digi2 Pro"},
	{ID: "hifiberry-studio-digi", Label: "HiFiBerry Studio Digi / AES"},
	{ID: "hifiberry-amp", Label: "HiFiBerry Amp / Amp+"},
	{ID: "hifiberry-amp100", Label: "HiFiBerry Amp100"},
	{ID: "hifiberry-amp3", Label: "HiFiBerry Amp3"},
	{ID: "hifiberry-amp4pro", Label: "HiFiBerry Amp4 Pro"},
	// IQaudIO
	{ID: "iqaudio-dac", Label: "IQaudIO Pi-DAC / Pi-DAC Zero"},
	{ID: "iqaudio-dacplus", Label: "IQaudIO Pi-DAC+ / Pi-DAC Pro / Pi-DigiAMP+"},
	{ID: "iqaudio-digi-wm8804-audio", Label: "IQaudIO Pi-Digi+"},
	{ID: "iqaudio-codec", Label: "IQaudIO Pi-Codec+ / Codec Zero"},
	{ID: "akkordion-iqdacplus", Label: "Digital Dreamtime Akkordion (IQaudIO DAC+ based)"},
	// Raspberry Pi branded
	{ID: "rpi-dacplus", Label: "Raspberry Pi DAC+"},
	{ID: "rpi-dacpro", Label: "Raspberry Pi DAC Pro"},
	{ID: "rpi-digiampplus", Label: "Raspberry Pi DigiAMP+"},
	{ID: "rpi-codeczero", Label: "Raspberry Pi Codec Zero"},
	// Allo
	{ID: "allo-boss-dac-pcm512x-audio", Label: "Allo Boss DAC"},
	{ID: "allo-boss2-dac-audio", Label: "Allo Boss2 DAC"},
	{ID: "allo-piano-dac-pcm512x-audio", Label: "Allo Piano DAC 2.0 (2.1 in stereo only)"},
	{ID: "allo-piano-dac-plus-pcm512x-audio", Label: "Allo Piano DAC 2.1"},
	{ID: "allo-digione", Label: "Allo DigiOne"},
	{ID: "allo-katana-dac-audio", Label: "Allo Katana DAC"},
	// JustBoom
	{ID: "justboom-dac", Label: "JustBoom DAC HAT / Amp HAT / DAC Zero / Amp Zero"},
	{ID: "justboom-digi", Label: "JustBoom Digi HAT / Digi Zero"},
	{ID: "justboom-both", Label: "JustBoom DAC + Digi (stacked)"},
	// PiFi
	{ID: "pifi-dac-hd", Label: "PiFi DAC HD"},
	{ID: "pifi-dac-zero", Label: "PiFi DAC Zero"},
	{ID: "pifi-40", Label: "PiFi 40W stereo amplifier"},
	{ID: "pifi-mini-210", Label: "PiFi Mini stereo amplifier"},
	// Dion Audio
	{ID: "dionaudio-loco", Label: "Dion Audio LOCO DAC-AMP"},
	{ID: "dionaudio-loco-v2", Label: "Dion Audio LOCO-V2 DAC-AMP"},
	{ID: "dionaudio-kiwi", Label: "Dion Audio KIWI Streamer"},
	// AudioInjector
	{ID: "audioinjector-wm8731-audio", Label: "AudioInjector Zero / Stereo"},
	{ID: "audioinjector-addons", Label: "AudioInjector Octo"},
	{ID: "audioinjector-ultra", Label: "AudioInjector Ultra"},
	{ID: "audioinjector-isolated-soundcard", Label: "AudioInjector Isolated"},
	{ID: "audioinjector-bare-i2s", Label: "AudioInjector bare I2S"},
	// Blokas
	{ID: "pisound", Label: "Blokas Pisound"},
	{ID: "pisound-pi5", Label: "Blokas Pisound (Pi 5)"},
	{ID: "pisound-micro", Label: "Blokas Pisound Micro"},
	// Others
	{ID: "applepi-dac", Label: "Orchard Audio ApplePi-DAC"},
	{ID: "i-sabre-q2m", Label: "Audiophonics I-Sabre Q2M"},
	{ID: "rra-digidac1-wm8741-audio", Label: "Red Rocks Audio DigiDAC1"},
	{ID: "dacberry400", Label: "DACBerry 400"},
	{ID: "chipdip-dac", Label: "Chip Dip DAC"},
	{ID: "interludeaudio-analog", Label: "Interlude Audio Analog HAT"},
	{ID: "interludeaudio-digital", Label: "Interlude Audio Digital HAT"},
	{ID: "cirrus-wm5102", Label: "Cirrus Logic Audio Card"},
	{ID: "fe-pi-audio", Label: "Fe-Pi Audio"},
	{ID: "superaudioboard", Label: "SuperAudioBoard"},
	{ID: "proto-codec", Label: "PROTO Audio Codec (WM8731)"},
	{ID: "mbed-dac", Label: "mbed AudioCODEC (TLV320AIC23B)"},
	{ID: "max98357a", Label: "Maxim MAX98357A I2S amplifier"},
	{ID: "wm8960-soundcard", Label: "Waveshare WM8960 Audio HAT"},
	{ID: "googlevoicehat-soundcard", Label: "Google AIY Voice HAT"},
	{ID: "merus-amp", Label: "Infineon MERUS Audio Amp"},
	{ID: "ghost-amp", Label: "Ghost amplifier"},
	{ID: "audiosense-pi", Label: "AudioSense-Pi"},
	{ID: "pibell", Label: "PiBell"},
	{ID: "ezsound-6x8iso", Label: "ezsound-6x8 Pi5 multichannel soundcard"},
}

// ByID resolves a catalog entry.
func ByID(id string) (Entry, bool) {
	for _, e := range Catalog {
		if e.ID == id {
			return e, true
		}
	}
	return Entry{}, false
}

var (
	overlayRE = regexp.MustCompile(`^\s*dtoverlay\s*=\s*([^,\s]+)(?:,(.*))?\s*$`)
	audioRE   = regexp.MustCompile(`(?i)^\s*dtparam\s*=\s*audio\s*=\s*(on|off)\s*$`)
)

// IsAudioOverlay reports whether the catalog knows this overlay. Anything
// else is left alone: guessing from a name is not a claim we can make.
func IsAudioOverlay(name string) bool {
	e, ok := ByID(name)
	return ok && e.ID != Onboard
}

func isAudioLine(line string) bool {
	if m := overlayRE.FindStringSubmatch(line); m != nil && IsAudioOverlay(m[1]) {
		return true
	}
	return audioRE.MatchString(line)
}

// splitBlock returns (before, block, after); block is nil when there is no
// managed block, and a Begin without End swallows to EOF.
func splitBlock(lines []string) (before, block, after []string) {
	start := slices.Index(lines, Begin)
	if start < 0 {
		return lines, nil, nil
	}
	end := slices.Index(lines[start+1:], End)
	if end < 0 {
		end = len(lines) - 1
	} else {
		end += start + 1
	}
	return lines[:start], lines[start : end+1], lines[end+1:]
}

func entryForLines(lines []string) string {
	audioOn := false
	for _, line := range lines {
		if m := overlayRE.FindStringSubmatch(line); m != nil && IsAudioOverlay(m[1]) {
			return m[1]
		}
		if a := audioRE.FindStringSubmatch(line); a != nil && strings.EqualFold(a[1], "on") {
			audioOn = true
		}
	}
	if audioOn {
		return Onboard
	}
	return ""
}

type ParsedConfig struct {
	Current    string   // entry ID, "" when nothing recognisable is configured
	Managed    bool     // the odioctl block is present
	StrayLines []string // active audio lines outside the managed block
}

func Parse(text string) ParsedConfig {
	before, block, after := splitBlock(strings.Split(text, "\n"))
	outside := slices.Concat(before, after)
	var stray []string
	for _, line := range outside {
		if isAudioLine(line) {
			stray = append(stray, line)
		}
	}
	if block != nil {
		return ParsedConfig{Current: entryForLines(block), Managed: true, StrayLines: stray}
	}
	return ParsedConfig{Current: entryForLines(outside), Managed: false, StrayLines: stray}
}

func renderBlock(e Entry) []string {
	audio := "off"
	if e.ID == Onboard {
		audio = "on"
	}
	lines := []string{Begin, "[all]", overlayReset, "dtparam=audio=" + audio}
	if dto := e.OverlayLine(); dto != "" {
		lines = append(lines, dto)
	}
	return append(lines, End)
}

func stripTrailingBlank(lines []string) []string {
	for len(lines) > 0 && strings.TrimSpace(lines[len(lines)-1]) == "" {
		lines = lines[:len(lines)-1]
	}
	return lines
}

// Apply returns config.txt text with e selected. Idempotent.
func Apply(text string, e Entry) string {
	before, _, after := splitBlock(strings.Split(text, "\n"))
	var body []string
	for _, line := range slices.Concat(before, after) {
		if isAudioLine(line) {
			body = append(body, DisabledPrefix+line)
		} else {
			body = append(body, line)
		}
	}
	body = stripTrailingBlank(body)
	if len(body) > 0 {
		body = append(body, "")
	}
	body = append(body, renderBlock(e)...)
	return strings.Join(body, "\n") + "\n"
}

// Unapply returns config.txt text with the managed block removed and disabled
// lines restored.
func Unapply(text string) string {
	before, _, after := splitBlock(strings.Split(text, "\n"))
	var body []string
	for _, line := range slices.Concat(before, after) {
		body = append(body, strings.TrimPrefix(line, DisabledPrefix))
	}
	body = stripTrailingBlank(body)
	if len(body) == 0 {
		return ""
	}
	return strings.Join(body, "\n") + "\n"
}

// FindConfigTxt returns the first existing config.txt candidate, "".
func FindConfigTxt() string {
	for _, p := range ConfigCandidates {
		if isFile(p) {
			return p
		}
	}
	return ""
}

func isFile(path string) bool {
	fi, err := os.Stat(path)
	return err == nil && fi.Mode().IsRegular()
}

func ReadConfig(path string) (string, error) {
	b, err := os.ReadFile(path)
	return string(b), err
}

// WriteConfig keeps a one-time .odioctl.bak of the pristine file, then writes
// atomically.
func WriteConfig(path, text string) error {
	backup := path + ".odioctl.bak"
	if _, err := os.Stat(backup); os.IsNotExist(err) {
		if orig, err := os.ReadFile(path); err == nil {
			if err := os.WriteFile(backup, orig, 0o644); err != nil {
				slog.Warn("could not write backup", "path", backup, "err", err)
			}
		}
	}
	return fsutil.AtomicWriteText(path, text)
}

func MarkRebootRequired() {
	if err := os.MkdirAll(filepath.Dir(RebootFlag), 0o755); err != nil {
		return
	}
	_ = os.WriteFile(RebootFlag, []byte("dac\n"), 0o644)
}

func RebootRequired() bool {
	_, err := os.Stat(RebootFlag)
	return err == nil
}

// Status is what the CLI and the web share about the current configuration.
type Status struct {
	Supported      bool     `json:"supported"`
	Config         string   `json:"config"`
	Current        string   `json:"current"`
	Managed        bool     `json:"managed"`
	StrayLines     []string `json:"stray_lines"`
	RebootRequired bool     `json:"reboot_required"`
}

// GetStatus inspects configPath ("" = auto-detect).
func GetStatus(configPath string) Status {
	path := configPath
	if path == "" {
		path = FindConfigTxt()
	}
	s := Status{Config: path, StrayLines: []string{}, RebootRequired: RebootRequired()}
	if path == "" {
		return s
	}
	text, err := ReadConfig(path)
	if err != nil {
		return s
	}
	p := Parse(text)
	s.Supported = true
	s.Current = p.Current
	s.Managed = p.Managed
	if p.StrayLines != nil {
		s.StrayLines = p.StrayLines
	}
	return s
}
